sys/include/event.h

```diff
@@ -0,0 +1,218 @@
+ *
+ * At the core, event_wait() uses thread flags to implement waiting for events
+ * to be queued. Thus event queues can be used safely and efficiently in combination
+ * with thread flags and msg queues.
+ *
+ * Examples:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * // simple event handler
+ * static void handler(event_t *event)
+ * {
+ *    printf("triggered 0x%08x\n", (unsigned)event);
+ * }
+ *
+ * static event_t event = { .handler=handler };
+ * static event_queue_t queue;
+ *
+ * int main(void)
+ * {
+ *     event_queue_init(&queue);
+ *     event_loop(&queue);
+ * }
+ *
+ * [...] event_post(&queue, &event);
+ *
+ * // example for event extended event struct
+ * typedef struct {
+ *     event_t super;
+ *     const char *text;
+ * } custom_event_t;
+ *
+ * static void custom_handler(event_t *event)
+ * {
+ *     custom_event_t *custom_event = (custom_event_t *)event;
+ *     printf("triggered custom event with text: \"%s\"\n", custom_event->text);
+ * }
+ *
+ * static custom_event_t custom_event = { .super.callback=custom_handler, .text="CUSTOM EVENT" };
+ *
+ * [...] event_post(&queue, &custom_event)
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief       Event API
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef EVENT_H
+#define EVENT_H
+
+#include <stdint.h>
+
+#include "irq.h"
+#include "thread_flags.h"
+#include "clist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef THREAD_FLAG_EVENT
+/**
+ * @brief   Thread flag use to notify available events in an event queue
+ */
+#define THREAD_FLAG_EVENT   (0x1)
+#endif
+
+/**
+ * @brief   event_queue_t static initializer
+ */
+#define EVENT_QUEUE_INIT    { .waiter=(thread_t *)sched_active_thread }
+
+/**
+ * @brief   event structure forward declaration
+ */
+typedef struct event event_t;
+
+/**
+ * @brief   event handler type definition
+ */
+typedef void(*event_handler_t)(event_t*);
+
+/**
+ * @brief   event structure
+ */
+struct event {
+    clist_node_t list_node;     /**< event queue list entry             */
+    event_handler_t handler;    /**< pointer to event handler function  */
+};
+
+/**
+ * @brief   event queue structure
+ */
+typedef struct {
+    clist_node_t event_list;    /**< list of queued events              */
+    thread_t *waiter;           /**< thread ownning event queue         */
+} event_queue_t;
+
+/**
+ * @brief   Initialize an event queue
+ *
+ * This will set the calling thread as owner of @p queue.
+ *
+ * @param[out]  queue   event queue object to initialize
+ */
+void event_queue_init(event_queue_t *queue);
+
+/**
+ * @brief   Queue an event
+ *
+ * @param[in]   queue   event queue to queue event in
+ * @param[in]   event   event to queue in event queue
+ */
+void event_post(event_queue_t *queue, event_t *event);
+
+/**
+ * @brief   Cancel a queued event
+ *
+ * This will remove a queued event from an event queue.
+ *
+ * @note: due to the underlying list implementation, this will run in O(n).
```

Comment:
remove the `:` please. It parses weirdly in Doxygen

![note with colon](https://user-images.githubusercontent.com/675644/32442139-a2f0b920-c2fa-11e7-8a69-cfad4055b390.png)

Reply:
fixed